lib/core/test/unit/Cardano/Wallet/DB/Arbitrary.hs

```diff
@@ -160,155 +159,166 @@ import qualified Data.ByteString as BS
-        pure $ KeyValPairs pairs
+        pure $ KeyValPairs $ L.sortOn fst pairs
 
-instance Arbitrary (PrimaryKey WalletId) where
-    shrink _ = []
+-- | For checkpoints, we make sure to generate them in order.
+instance {-# OVERLAPS #-} (Arbitrary k, Ord k, GenState s)
+    => Arbitrary (KeyValPairs k (ShowFmt (Wallet s DummyTarget))) where
+    shrink = genericShrink
     arbitrary = do
-        bytes <- B8.pack . pure <$> elements ['a'..'k']
-        return $ PrimaryKey $ WalletId $ hash bytes
+        pairs <- choose (1, 10) >>= flip vectorOf arbitrary
+        pure $ KeyValPairs
+            $ fmap (second ShowFmt)
+            $ L.sortOn (\(k,cp) -> (k, (view #slotId $ currentTip cp)))
+            $ pairs
 
-instance GenState s => Arbitrary (Wallet s DummyTarget) where
-    shrink w =
-        [ unsafeInitWallet u (currentTip w) s genesisParameters
-        | (u, s) <- shrink (utxo w, getState w) ]
-    arbitrary = unsafeInitWallet
-        <$> arbitrary
-        <*> pure (BlockHeader (SlotId 0 0) (Quantity 0) (Hash "hash"))
-        <*> arbitrary
-        <*> pure genesisParameters
+instance Arbitrary GenTxHistory where
+    shrink (GenTxHistory h) = map GenTxHistory (shrinkList shrinkOneTx h)
+      where
+        shrinkOneTx
+            :: (DummyTarget.Tx, TxMeta)
+            -> [(DummyTarget.Tx, TxMeta)]
+        shrinkOneTx (tx, meta) =
+            [(tx', meta) | tx' <- shrink tx]
 
-instance Arbitrary Address where
-    -- No Shrinking
-    arbitrary = oneof
-        [ pure $ Address "ADDR01"
-        , pure $ Address "ADDR02"
-        , pure $ Address "ADDR03"
-        , pure $ Address "ADDR04"
-        , pure $ Address "ADDR05"
-        , pure $ Address "ADDR06"
-        , pure $ Address "ADDR07"
-        , pure $ Address "ADDR08"
-        , pure $ Address "ADDR09"
-        , pure $ Address "ADDR10"
-        ]
+    -- Ensure unique transaction IDs within a given batch of transactions to add
+    -- to the history.
+    arbitrary = GenTxHistory . sortTxHistory <$> do
+        -- NOTE
+        -- We discard pending transaction from any 'GenTxHistory since,
+        -- inserting a pending transaction actually has an effect on the
+        -- checkpoint's pending transactions of the same wallet.
+        filter (not . isPending . snd) <$> arbitrary
+      where
+        sortTxHistory = filterTxHistory @DummyTarget Descending wholeRange
 
-instance Arbitrary (SeqState DummyTarget) where
-    shrink (SeqState intPool extPool ixs) =
-        (\(i, e, x) -> SeqState i e x) <$> shrink (intPool, extPool, ixs)
+instance Arbitrary MockChain where
+    shrink (MockChain chain) =
+        [ MockChain chain'
+        | chain' <- shrinkList shrinkBlock chain
+        , not (null chain')
+        ]
+      where
+        shrinkBlock (Block h txs) = Block h <$> shrinkList shrink txs
     arbitrary = do
-        SeqState <$> arbitrary <*> arbitrary <*> arbitrary
+        n0 <- choose (1, 10)
```

Comment:
the same comment/question here

Reply:
Indeed. It has proven to be sufficient long chain already to find quite a lot of bugs :sweat_smile: